broadinstitute/scp-ingest-pipeline
# Rank genes in study by relevance (SCP-5399)

This enables ranking genes in a study by various dimensions of relevance.  That helps prioritize expensive caching.

### Overview
The rank-genes pipeline sorts genes using the study's publication, differential expression, and global interest metrics.  For example, the NeMO study "Neuronal diversity in the human hypothalamus" ([SCP2247](https://singlecell.broadinstitute.org/single_cell/study/SCP2247/neuronal-diversity-in-the-human-hypothalamus)) has [a scientific journal publication](https://www.science.org/doi/10.1126/sciadv.adf6251) and DE metrics for its cell type-like annotation.  Global interest metrics, like Wikipedia page views and PubMed citation counts -- are also available for each gene.  

### Output, prior work
The top 100 ranked genes are output in a TSV file, and uploaded to a bucket.  These top genes will be used to prioritize the image caching pipeline.  The output is currently fairly rich, as this rank-genes pipeline derived from an earlier prototype of a gene leads pipeline.  It can be economized or extended in future work if needed.

### Limitations
The current pipeline works with one of the NeMO exemplars, but might need updates for wider use.  For example, small updates might be needed if the study lacks a publication, DE, or both.  

### Incidental cleanup
To simplify testing, I removed code under `ingest/genomes/` and related tests.  That was an exploratory branch of work from several years ago, but never actually got used.  See e.g. #129 for context on that code.

I also removed a large `try:` import block.  As noted in code comments, this was an effective but very unwieldy technique to enable importing Ingest Pipeline as a package.  That package offering has also been unused and unmaintained for several years, so we can opportunistically remove those try-import blocks to eliminate lots of boilerplate code.

### Test
New automated tests verify this new pipeline.  To manually test it:

1.  Ensure your Ingest Pipeline environment is set up, e.g. `echo $DATABASE_NAME` should output `single_cell_portal_development`, activate your virtual environment, etc.
2. Copy the 155K-cell NeMO exemplar study, [SCP2247](https://singlecell.broadinstitute.org/single_cell/study/SCP2247/neuronal-diversity-in-the-human-hypothalamus), to your local SCP instance if you haven't already, and fully initialize the study, including DE
3. Get the study accession of your local copy of production SCP2247
4. Run the following; note that `--study-id` and `--study-file-id` are incidentally-needed placeholders:

> python3 ingest/ingest_pipeline.py --study-id 5d276a50421aa9117c982845 --study-file-id 5dd5ae25421aa910a723a337 rank_genes --study-accession (your study accession) --publication="https://doi.org/10.1126/sciadv.adf6251" --rank-genes

5.  Go to your study's GCP bucket
6. Confirm output file exists at `_scp_internal/ranked_genes/ranked_genes.tsv`

### More context
Further detail is available in the 2023-12-19 SCP demo recording ([overview](https://drive.google.com/file/d/1WK-MAM84nZUb6o_ynHnf0XCI4X0kKJii/view?t=15), [technical notes](https://drive.google.com/file/d/1WK-MAM84nZUb6o_ynHnf0XCI4X0kKJii/view?t=10m7s)).  This relates to SCP-5399.
